isogeny_primes.py

```diff
@@ -793,6 +793,93 @@ def get_aux_primes(K, norm_bound, C_K, h_K, contains_imaginary_quadratic):
     return aux_primes
 
 
+def filter_possible_values(possible_values_list,fq,prime_field):
+
+    output = []
+
+    for c in possible_values_list:
+        if c ** 2 == prime_field(1):
+            output.append(c)
+        elif c ** 2 == prime_field(fq.norm()):
+            output.append(c)
+        else:
+            fq_char = ZZ(fq.norm()).prime_divisors()[0]
+            possible_mid_coeffs = [a for a in range(-fq.norm(),fq.norm()+1) if prime_field(a) == c + prime_field(fq.norm()/c)  ]
+            possible_weil_polys = [x ** 2 + a * x + fq.norm() for a in possible_mid_coeffs]
+            elliptic_weil_polys = [f for f in possible_weil_polys if weil_polynomial_is_elliptic(f,fq_char,fq.residue_class_degree())]
+            if elliptic_weil_polys:
+                output.append(c)
+    return output
+
+def get_possible_vals_at_gens(gens_info, eps, embeddings, residue_field, prime_field):
+
+    output = {}
+    # frak_p0 = K.primes_above(p)[0]  # choice of p_0
+    # residue_field = frak_p0.residue_field(names='z')
+    # prime_field = GF(p)
+
+    for class_gp_gen in gens_info:
+        class_gp_order, alpha = gens_info[class_gp_gen]
+        alpha_to_eps = eps_exp(alpha, eps, embeddings)
+        alpha_to_eps_mod_p0 = residue_field(alpha_to_eps)
+
+        try:
+            c_power_12h = prime_field(alpha_to_eps_mod_p0)
+            possible_values = c_power_12h.nth_root(12*class_gp_order, all=True)
+            filtered_values = filter_possible_values(possible_values)
+            output[class_gp_gen] = filtered_values
+        except TypeError:
+            # means alpha_to_eps_mod_p0 is not in GF(p) so can ignore p and move on
+            return {}
+
+    return output
+
+
+def tuple_exp(tup,exp_tup):
+    return tuple((t**e for t,e in zip(tup,exp_tup)))
+
+
+def final_filter(Kgal, aux_primes, my_gens_ideals, gens_info, p, eps, embeddings):
+
+    frak_p0 = Kgal.primes_above(p)[0]  # choice of p_0
+    residue_field = frak_p0.residue_field(names='z')
+    prime_field = GF(p)
+
+    # Step 1
+    possible_vals_at_gens = get_possible_vals_at_gens(gens_info, eps, embeddings, residue_field, prime_field)
+
+    if not possible_vals_at_gens:
+        return False
+
+    # Step 2
+
+    # a list of tuples
+    possible_vals_cart_prod = list(product(*[possible_vals_at_gens[q] for q in my_gens_ideals]))
+
+    # Step 3
+    for q in aux_primes:
+        exponents_in_class_group = q.exponents()
```

Comment:
Why doesn't q.exponents() raise an error here? From reading the code have the feeling that q is just a fractional ideal here right?

```
K.<a> = QQ.extension(x^3 - 4095)                                          
sage: G = K.class_group()                                                       
sage: (K*29).factor()[0][0].exponents() 
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
.... snip
AttributeError: 'NumberFieldFractionalIdeal' object has no attribute 'exponents'
```

While if we first make it an element of the class group:

```
sage: G((K*29).factor()[0][0]).exponents()                                      
(11, 5, 2)
```